internal/bufferedwrites/buffered_write_handler.go

```diff
@@ -0,0 +1,127 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package bufferedwrites
+
+import (
+	"fmt"
+	"math"
+	"time"
+
+	"github.com/googlecloudplatform/gcsfuse/v2/internal/block"
+	"golang.org/x/sync/semaphore"
+)
+
+// Note: All the write operations take inode lock in fs.go, hence we don't need any locks here
+// as we will get write operations serially.
+
+// BufferedWriteHandler is responsible for filling up the buffers with the data
+// as it receives and handing over to uploadHandler which uploads to GCS.
+type BufferedWriteHandler struct {
+	current   block.Block
+	blockPool *block.BlockPool
+	// Total size of data buffered so far. Some part of buffered data might have
+	// been uploaded to GCS as well.
+	totalSize int64
+	// Stores the mtime value updated by kernel as part of setInodeAttributes call.
+	mtime time.Time
+}
+
+// WriteFileInfo is used as part of serving fileInode attributes (GetInodeAttributes call).
+type WriteFileInfo struct {
+	TotalSize int64
+	Mtime     time.Time
+}
+
+// InitBWHandler initialises the bufferedWriteHandler struct.
+func InitBWHandler(blockSize int64, maxBlocks int32, globalMaxBlocksSem *semaphore.Weighted) (bwh *BufferedWriteHandler, err error) {
+	bp, err := block.InitBlockPool(blockSize, maxBlocks, globalMaxBlocksSem)
+	if err != nil {
+		return
+	}
+
+	bwh = &BufferedWriteHandler{
+		current:   nil,
+		blockPool: bp,
+		totalSize: 0,
+		mtime:     time.Now(),
+	}
+	return
+}
+
+// Write writes the given data to the buffer. It writes to an existing buffer if
+// the capacity is available otherwise writes to a new buffer.
+func (wh *BufferedWriteHandler) Write(data []byte, offset int64) (err error) {
+	if offset > wh.totalSize {
+		// TODO: Will be handled as part of ordered writes.
+		return fmt.Errorf("non sequential writes")
+	}
+
+	dataWritten := 0
+	for dataWritten < len(data) {
+		if wh.current == nil {
+			wh.current, err = wh.blockPool.Get()
+			if err != nil {
+				return
+			}
+		}
+
+		remainingBlockSize := float64(wh.blockPool.BlockSize()) - float64(wh.current.Size())
+		pendingDataForWrite := float64(len(data)) - float64(dataWritten)
+		bytesToCopy := int(math.Min(remainingBlockSize, pendingDataForWrite))
+		err := wh.current.Write(data[dataWritten : dataWritten+bytesToCopy])
+		if err != nil {
+			return err
+		}
+
+		dataWritten += bytesToCopy
+
+		if wh.current.Size() == wh.blockPool.BlockSize() {
+			// TODO: err = trigger upload
+			if err != nil {
+				return err
+			}
+			wh.current = nil
+		}
+	}
+
+	wh.totalSize += int64(dataWritten)
+	return
+}
+
+// Sync uploads all the pending full buffers to GCS.
+func (wh *BufferedWriteHandler) Sync() (err error) {
+	// TODO: Will be added after uploadHandler changes are done.
+	return nil
```

Comment:
nit:

```suggestion
	return fmt.Errorf("not implemented")
```

And in other unimplemented function below.